src/realm/mixed.cpp

```diff
@@ -85,88 +85,207 @@ inline int compare_float(Float a_raw, Float b_raw)
+    if (std::isnan(rhs))
         return 1;
-    else if (m_type < b.m_type)
-        return -1;
 
-    // Now we are sure the two types are the same
+    // Ints with magnitude <= 2**53 can be precisely represented as doubles.
+    // Additionally, doubles outside of this range can't have a fractional component.
+    static const int64_t kEndOfPreciseDoubles = 1ll << 53;
+    if (lhs <= kEndOfPreciseDoubles && lhs >= -kEndOfPreciseDoubles) {
+        return compare_float(double(lhs), rhs);
+    }
+
+    // Large magnitude doubles (including +/- Inf) are strictly > or < all Longs.
+    static const double kBoundOfLongRange = -static_cast<double>(LLONG_MIN); // positive 2**63
+    if (rhs >= kBoundOfLongRange)
+        return -1; // Can't be represented in a Long.
+    if (rhs < -kBoundOfLongRange)
+        return 1; // Can be represented in a Long.
+
+    // Remaining Doubles can have their integer component precisely represented as long longs.
+    // If they have a fractional component, they must be strictly > or < lhs even after
+    // truncation of the fractional component since low-magnitude lhs were handled above.
+    return compare_generic(lhs, int64_t(rhs));
+}
+} // namespace _impl
+
+bool Mixed::is_comparable(const Mixed& l, const Mixed& r)
+{
+    if (l.is_null() || r.is_null())
+        return true;
+
+    DataType l_type = l.get_type();
+    DataType r_type = r.get_type();
+    if (l_type == r_type)
+        return true;
+    bool l_is_numeric = l_type == type_Int || l_type == type_Bool || l_type == type_Float || l_type == type_Double ||
+                        l_type == type_Decimal;
+    bool r_is_numeric = r_type == type_Int || r_type == type_Bool || r_type == type_Float || r_type == type_Double ||
+                        r_type == type_Decimal;
+    if (l_is_numeric && r_is_numeric) {
+        return true;
+    }
+    if ((l_type == type_String && r_type == type_Binary) || (r_type == type_String && l_type == type_Binary)) {
+        return true;
+    }
+    return false;
+}
+
+
+int Mixed::compare(const Mixed& b) const
+{
     if (is_null()) {
-        // Both are null
-        return 0;
+        return b.is_null() ? 0 : -1;
     }
+    if (b.is_null())
+        return 1;
 
+    // None is null
     switch (get_type()) {
+        case type_Bool: {
+            int64_t i_val = bool_val ? 1 : 0;
+            switch (b.get_type()) {
+                case type_Int:
+                    return _impl::compare_generic(i_val, b.int_val);
+                case type_Bool:
+                    return _impl::compare_generic(bool_val, b.bool_val);
+                case type_Float:
+                    return _impl::compare_long_to_double(i_val, b.float_val);
+                case type_Double:
+                    return _impl::compare_long_to_double(i_val, b.double_val);
+                case type_Decimal:
+                    return _impl::compare_decimals(Decimal128(i_val), b.decimal_val);
+                default:
```

Comment:
Should we not add all the remaining types here so we get a compile time error when we add an additional type? same all over)

Reply:
It will just be annoying. It is only when we add another numerical data type that we should modify this.